contrib/go/src/python/pants/contrib/go/subsystems/fetcher.py

```diff
@@ -76,8 +76,34 @@ def root(self):
         imported_repo = self._meta_tag_reader.get_imported_repo(self.import_path)
         if imported_repo:
             return imported_repo.import_prefix
+
+        # The import path is for an internal package which is described here:
+        #   https://golang.org/cmd/go/#hdr-Internal_Directories.
+        #
+        # Since an internal directory is only importable in the tree rooted at its parent, we
+        # assume its parent must have a public meta tag page. Failing that, we crawl all the way
+        # back up the ancestry.
+        import_path_components = self.import_path.split("/")
+        try:
+            internal_index = import_path_components.index("internal")
+        except ValueError:
+            internal_import_public_parent_candidates = []
         else:
-            raise FetchError(f'No <meta name="go-import"> tag found at {self.import_path}')
+            internal_import_public_parent_candidates = import_path_components[:internal_index]
+
+        while internal_import_public_parent_candidates:
+            parent_import_path = "/".join(internal_import_public_parent_candidates)
+            imported_repo = self._meta_tag_reader.get_imported_repo(parent_import_path)
+            if imported_repo:
+                logger.info(
+                    "Found public root for private internal import {} in parent {}.".format(
+                        self.import_path, parent_import_path
+                    )
+                )
+                return imported_repo.import_prefix
+            internal_import_public_parent_candidates.pop()
```

Comment:
Maybe add a comment here that this is what walks up to the ancestor? Reading earlier in this loop confused me a bit until I saw this.

Reply:
Thanks for taking a look. I hate to say I'm not going to do this, but with a 22 shard tottering CI for pantsbuild's oldest maintained branch that I just had to pay Travis to even be able to run at all - I'm going to save the CI credits for this one. On the off chance we do need to go back and do maintenance on this bit I think your code review comment will be enough to help out a spelunker that git blames and follows the breadcrumb.